From 9ce8d13cbec2e4ccdc68c435ad0d460a26aaea9f Mon Sep 17 00:00:00 2001
From: Avi Niddam <aniddam@habana.ai>
Date: Thu, 05 Jun 2025 11:54:50 +0300
Subject: [PATCH] [SW-228856] fit: prepare no duplicate

We encountered several code duplications related to
fit_foreach_property.
So, we want to create a new fit services infrastructure
to solve this issue in a clean way.
The main Lego pieces are the newly created files:
fit_services.c/h
This infra is being prepared and in this commit we create
_fit_foreach_property which is a temp function to pass CI.
General remark:
This infrastructure is intended to support ANY of the
fit related functions that we find today in fit_utils.c
or fit.c
Therefore, from now, please, anyone doing a change related
to the fit files should carefully double-check:
Any unified code can be (and should be) located in
fit_services.c/h
For this, please, follow the example of how it has been
done with fit_foreach_property.

Change-Id: Id62b989794fd99f817113d7fe941b06d3eb792cf
Signed-off-by: Avi Niddam <aniddam@habana.ai>
---

diff --git a/include/fit_services.h b/include/fit_services.h
new file mode 100644
index 0000000..266353c
--- /dev/null
+++ b/include/fit_services.h
@@ -0,0 +1,21 @@
+/* SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (C) 2025 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#ifndef _EMB_CMN_MOD_FIT_SERVICES_H_
+#define _EMB_CMN_MOD_FIT_SERVICES_H_
+
+#include <stdint.h>
+#include "libfdt.h"
+#include "fdt.h"
+
+typedef int (*_fit_property_match_cb)(const void *fit,
+		const struct fdt_property *prop, void *data);
+
+int _fit_foreach_property(const void *fit, const char *node_path,
+			_fit_property_match_cb match_cb, void *data);
+
+#endif
diff --git a/lib/common/fit_services.c b/lib/common/fit_services.c
new file mode 100644
index 0000000..23104a4
--- /dev/null
+++ b/lib/common/fit_services.c
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2025 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#include <stdint.h>
+#include <hl_fit_if.h>
+#include <libfdt.h>
+
+#include <fit_services.h>
+#include <emb_cmn_log.h>
+
+int _fit_foreach_property(const void *fit, const char *node_path,
+			_fit_property_match_cb match_cb, void *data)
+{
+	int ret, count = 0;
+	int noffset, property;
+	const struct fdt_property *prop;
+
+	noffset = fdt_path_offset(fit, node_path);
+	if (noffset < 0) {
+		EMB_LOG_RAW_ERR("Cannot find entry ");
+		EMB_LOG_RAW(node_path);
+		EMB_LOG_RAW("\n\r");
+		return noffset;
+	}
+
+	fdt_for_each_property_offset(property, fit, noffset) {
+		prop = fdt_get_property_by_offset(fit, property, NULL);
+		if (!prop)
+			continue;
+
+		ret = match_cb(fit, prop, data);
+		if (!ret)
+			count++;
+	}
+
+	return count;
+}
